File: 19-Iterator/src/go/main.go
package main

import "fmt"

// printAll works with ANY iterator — the whole point of the pattern.
func printAll(it Iterator) {
	for it.HasNext() {
		fmt.Printf("  %s\n", it.Next())
	}
}

func main() {
	fmt.Println("=== Iterator Pattern Demo (Go) ===")
	fmt.Println()

	playlist := &Playlist{}
	playlist.AddSong(NewSong("Bohemian Rhapsody", "Queen"))
	playlist.AddSong(NewSong("Hotel California", "Eagles"))
	playlist.AddSong(NewSong("Stairway to Heaven", "Led Zeppelin"))

	fmt.Println("Forward:")
	printAll(playlist.Iterator())

	fmt.Println("\nReverse:")
	printAll(playlist.ReverseIterator())

	fmt.Println("\nTwo independent cursors:")
	a := playlist.Iterator()
	b := playlist.Iterator()
	fmt.Printf("  a.Next(): %s\n", a.Next())
	fmt.Printf("  a.Next(): %s\n", a.Next())
	fmt.Printf("  b.Next(): %s   (b is unaffected by a)\n", b.Next())
}

File: 19-Iterator/src/cpp/main.cpp
#include <iostream>
#include <memory>
#include "Song.h"
#include "Iterator.h"
#include "Playlist.h"

// Works with ANY iterator — the whole point of the pattern.
void print_all(std::unique_ptr<Iterator> it) {
    while (it->has_next()) {
        std::cout << "  " << it->next().to_string() << "\n";
    }
}

int main() {
    std::cout << "=== Iterator Pattern Demo (C++) ===\n\n";

    Playlist playlist;
    playlist.add_song(Song("Bohemian Rhapsody", "Queen"));
    playlist.add_song(Song("Hotel California", "Eagles"));
    playlist.add_song(Song("Stairway to Heaven", "Led Zeppelin"));

    std::cout << "Forward:\n";
    print_all(playlist.iterator());

    std::cout << "\nReverse:\n";
    print_all(playlist.reverse_iterator());

    std::cout << "\nTwo independent cursors:\n";
    auto a = playlist.iterator();
    auto b = playlist.iterator();
    std::cout << "  a.next(): " << a->next().to_string() << "\n";
    std::cout << "  a.next(): " << a->next().to_string() << "\n";
    std::cout << "  b.next(): " << b->next().to_string() << "   (b is unaffected by a)\n";

    return 0;
}
